### ISSUE TITLE
Incorrect behavior of System.IO methods

### ISSUE THREAD

**Issue by Swaglis** [dotnet/corefx#41526]
# Path.GetFileNameWithoutExtension does not throw

After upgrading to .NET Core 3.0 I noticed that when passing an path containing invalid characters (e.g. "file|name.txt") the method no longer throws ArgumentException as it did before. According to the [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.getfilenamewithoutextension?view=netcore-3.0), this should still be the case.
Same applies to Directory.EnumerateDirectories method, in which not only the path is not checked and the method throws IO exception instead, neither the search patter argument is verified.

**Comment by carlossanlop** [comment]
I see what you mean:
- `Directory.EnumerateDirectories` throws `ArgumentException` if either the `directory` or the `expression` arguments specifically contain a null character `\0`. This is the only invalid character we consider for this method. Passing a `|` character will not throw `ArgumentException`, it will instead throw:
    ```
    System.IO.IOException: 'The filename, directory name, or volume label syntax is incorrect. 
    ```
- `Path.GetFileNameWithoutExtension` throws nothing if you pass either `|` or `\0`.

Neither of these methods call `GetInvalidPathChars` or `GetInvalidFileNameChars` at any point.

@JeremyKuhne should we update the documentation (intended behavior) or is it a bug that we don't throw when an invalid character is passed?

**Comment by scalablecory** [comment]
@carlossanlop have we verified that behavior changed since .NET Core 2.x or Framework?

Presumably the current behavior is that `IOException` is used when OS validates path and `ArgumentException` is used when the method validates path. If this was always the case, and we just loosened our validation and relegated more to .NET, then it wouldn't be a breaking change.

**Comment by JeremyKuhne** [comment]
Yes, the documentation needs updated.

We have explicitly removed all path character validity checks in Core 2.1 (outside of embedded nulls). Our guessing was incorrect, particularly with mounted directories from network shares and Unix file systems. Additionally doing so was pretty expensive with very little benefit.

I did a write-up that goes into more detail about what we changed in 2.1.  https://blogs.msdn.microsoft.com/jeremykuhne/2018/03/08/system-io-in-net-core-2-1-sneak-peek/
